compiler/src/abi/elements.rs

```diff
@@ -0,0 +1,272 @@
+    }
+}
+
+/// All public interfaces of Vyper contract.
+#[derive(Debug, PartialEq, Clone)]
+pub struct Contract {
+    /// All events defined in a contract.
+    pub events: Vec<Event>,
+    /// All public functions defined in a contract.
+    pub functions: Vec<Function>,
+}
+
+impl Contract {
+    /// Serialize the contract into a valid JSON ABI.
+    pub fn json(&self) -> Result<String, CompileError> {
+        serde_json::to_string(self)
+            .map_err(|_| CompileError::static_str("Unable to serialize contract to json"))
+    }
+}
+
+impl Serialize for Contract {
+    fn serialize<S>(&self, serializer: S) -> Result<<S as Serializer>::Ok, <S as Serializer>::Error>
+    where
+        S: Serializer,
+    {
+        let mut seq = serializer.serialize_seq(None)?;
+
+        for event in self.events.iter() {
+            seq.serialize_element(event)?;
+        }
+
+        for function in self.functions.iter() {
+            seq.serialize_element(function)?;
+        }
+
+        seq.end()
+    }
+}
+
+/// An event interface.
+#[derive(Serialize, Debug, PartialEq, Clone)]
+pub struct Event {
+    /// The event's name.
+    pub name: String,
+    // FIXME: This attribute should be added when serialized instead of being defined in the struct.
+    /// The type of an event (Always "event").
+    #[serde(rename = "type")]
+    pub typ: String,
+    /// All event fields.
+    #[serde(rename = "inputs")]
+    pub fields: Vec<EventField>,
+    /// True if the event was declared as anonymous.
+    pub anonymous: bool,
+}
+
+/// A single event field.
+#[derive(Serialize, Debug, PartialEq, Clone)]
+pub struct EventField {
+    /// The event field's name.
+    pub name: String,
+    /// The type of an event (e.g. uint256, address, bytes100,...)
+    #[serde(rename = "type")]
+    pub typ: VariableType,
+    /// True if the field is part of the log’s topics, false if it is one of the log’s data segment.
+    pub indexed: bool,
+}
+
+/// A function interface.
+#[derive(Serialize, Debug, PartialEq, Clone)]
+pub struct Function {
+    /// The function's name.
+    pub name: String,
+    /// The type of a function (Function, Constructor, Receive, and Fallback)
+    #[serde(rename = "type")]
+    pub typ: FuncType,
+    /// All function inputs.
+    pub inputs: Vec<FuncInput>,
+    /// All function outputs.
+    pub outputs: Vec<FuncOutput>,
+}
+
+/// A single function input.
+#[derive(Serialize, Debug, PartialEq, Clone)]
+pub struct FuncInput {
+    /// The input's name.
+    pub name: String,
+    /// The input's type.
+    #[serde(rename = "type")]
+    pub typ: VariableType,
+}
+
+/// A single function output.
+#[derive(Serialize, Debug, PartialEq, Clone)]
+pub struct FuncOutput {
+    /// The output's name.
+    pub name: String,
+    /// The output's type.
+    #[serde(rename = "type")]
+    pub typ: VariableType,
+}
+
+// FIXME: This should be removed in favor of separate structs for each function type.
```

Comment:
curious why?

Reply:
If the ABIs built for contracts are used internally by the compiler for things like module imports, it would be best to have data structures that are easy to use as opposed to 1-1 reflections of the serialized JSON format.

Reply:
ah gotcha. perfect!